// @chakra-ui/vue/dist/esm/CCheckboxGroup.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

import { R as createStyledAttrsMixin, M as isDef, J as useId, C as cleanChildren, A as cloneVNodeElement } from './index-2216d089.js';
import './chakra-ui-styled-system.esm-a8a2e91f.js';
import '@emotion/css';
import 'vue';
import './color-mode-observer-de5da461.js';
import CBox from './CBox.js';
import { a as SNA } from './props.types-cbb7b1ff.js';

/**
 * Hey! Welcome to @chakra-ui/vue CCheckboxGroup
 *
 * CCheckboxGroup component allows the user to group a
 * related collection of checkboxes in one visual region.
 *
 * @see Docs     https://vue.chakra-ui.com/checkbox
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CCheckboxGroup/CCheckboxGroup.js
 * @see A11y     https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CCheckbox/accessibility.md
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2/#button
 */
/**
 * CButtonGroup component
 *
 * Wrapper for children `CCheckbox` components
 *
 * @extends CBox
 * @see Docs https://vue.chakra-ui.com/button
 */

var CCheckboxGroup = {
  name: 'CCheckboxGroup',
  mixins: [createStyledAttrsMixin('CCheckboxGroup')],
  model: {
    prop: 'value',
    event: 'change'
  },
  props: {
    name: String,
    variantColor: String,
    size: String,
    defaultValue: Array,
    isInline: Boolean,
    value: Array,
    spacing: {
      type: SNA,
      default: 2
    }
  },
  data: function data() {
    return {
      values: this.defaultValue || []
    };
  },
  computed: {
    isControlled: function isControlled() {
      return isDef(this.value);
    },
    _values: function _values() {
      return this.isControlled ? this.value : this.values;
    },
    checkboxName: function checkboxName() {
      return this.name || "checkbox-group-" + useId();
    }
  },
  methods: {
    /**
     * Handles change event for checkbox group
     * @param {Event} event Event object
     */
    onChange: function onChange(val, event) {
      var ref = event.target;
      var checked = ref.checked;
      var value = ref.value;
      var newValues;

      if (checked) {
        newValues = this._values.concat([value]);
      } else {
        newValues = this._values.filter(function (val) {
          return val !== value;
        });
      }

      if (!this.isControlled) {
        this.values = newValues;
      }

      this.$emit('change', newValues);
    }
  },
  render: function render(h) {
    var this$1 = this;
    var children = cleanChildren(this.$slots.default);
    var clones = children.map(function (vnode, index) {
      var isLastCheckbox = children.length === index + 1;
      var spacingProps = this$1.isInline ? {
        mr: this$1.spacing
      } : {
        mb: this$1.spacing
      };
      var el = cloneVNodeElement(vnode, {
        props: {
          size: this$1.size,
          variantColor: this$1.variantColor,
          name: this$1.checkboxName + "-" + index,
          isChecked: this$1._values.includes(vnode.componentOptions.propsData.value)
        },
        on: {
          change: this$1.onChange
        }
      }, h);
      return h(CBox, {
        attrs: Object.assign({}, {
          display: this$1.isInline ? 'inline-block' : 'block'
        }, !isLastCheckbox && spacingProps)
      }, [el]);
    });
    return h('div', {
      class: [this.className],
      attrs: Object.assign({}, this.computedAttrs, {
        role: 'group'
      })
    }, clones);
  }
};

export default CCheckboxGroup;
