// @chakra-ui/vue/dist/esm/CHeading.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

import { R as createStyledAttrsMixin } from './index-2216d089.js';
import './chakra-ui-styled-system.esm-a8a2e91f.js';
import '@emotion/css';
import 'vue';
import './color-mode-observer-de5da461.js';
import { u as useTruncated } from './text.utils-6bf402a1.js';

/**
 * Hey! Welcome to @chakra-ui/vue Heading
 *
 * Headings are used for rendering headlines.
 *
 * Heading composes CBox so you can use all the style props
 * and add responsive styles as well. It renders an <h2> tag by default.
 *
 * @see Docs     https://vue.chakra-ui.com/heading
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CHeading/CHeading.js
 */
var sizes = {
  '2xl': ['4xl', null, '5xl'],
  xl: ['3xl', null, '4xl'],
  lg: ['xl', null, '2xl'],
  md: 'xl',
  sm: 'md',
  xs: 'sm'
};
/**
 * CHeading component
 *
 * The CHeading is used for rendering headlines.
 *
 * @extends CBox
 * @see Docs https://vue.chakra-ui.com/heading
 */

var CHeading = {
  name: 'CHeading',
  mixins: [createStyledAttrsMixin('CHeading')],
  props: {
    size: {
      type: [String, Array, Object],
      default: 'xl'
    },
    as: {
      type: String,
      default: 'h2'
    },
    isTruncated: Boolean
  },
  computed: {
    componentStyles: function componentStyles() {
      return Object.assign({}, {
        fontSize: sizes[this.size],
        lineHeight: 'shorter',
        fontWeight: 'bold',
        fontFamily: 'heading'
      }, this.isTruncated && useTruncated());
    }
  },
  render: function render(h) {
    return h(this.as, {
      class: this.className,
      attrs: this.computedAttrs
    }, this.$slots.default);
  }
};

export default CHeading;
