// @chakra-ui/vue/dist/esm/index-dbd7e06a.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

import { R as createStyledAttrsMixin } from './index-2216d089.js';
import { a as SNA } from './props.types-cbb7b1ff.js';

var flexProps = {
  as: {
    type: String,
    default: 'div'
  },
  align: SNA,
  justify: SNA,
  wrap: SNA,
  direction: SNA,
  size: SNA
};

/**
 * Hey! Welcome to @chakra-ui/vue Flex
 *
 * `CFlex` is `CBox` with `display: flex` and comes with
 * helpful style shorthands. It renders a `div` element.
 *
 * @see Docs     https://vue.chakra-ui.com/flex
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CFlex/CFlex.js
 */
/**
 * CFlex component
 *
 * `CFlex` is `CBox` with display: flex and comes with helpful style shorthands.
 *
 * @see Docs https://vue.chakra-ui.com/flex
 */

var CFlex = {
  name: 'CFlex',
  mixins: [createStyledAttrsMixin('CFlex')],
  props: flexProps,
  computed: {
    componentStyles: function componentStyles() {
      return {
        display: 'flex',
        flexDirection: this.direction,
        alignItems: this.align,
        justifyContent: this.justify,
        flexWrap: this.wrap,
        h: this.size,
        w: this.size
      };
    }
  },
  render: function render(h) {
    return h(this.as, {
      class: this.className,
      attrs: this.computedAttrs,
      on: this.computedListeners
    }, this.$slots.default);
  }
};

export { CFlex as C, flexProps as f };
