// @chakra-ui/vue/dist/cjs/badge.styles-5954ab1f.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

'use strict';

var utils = require('./index-aaaef82b.js');

var solidStyle = function solidStyle(ref) {
  var colors = ref.theme.colors;
  var color = ref.color;

  var _color = colors[color] && colors[color][500];

  var darkModebackgroundColor = utils.addOpacity(_color, 0.6);
  return {
    light: {
      backgroundColor: utils.get(color, 500),
      color: 'white'
    },
    dark: {
      backgroundColor: darkModebackgroundColor,
      color: 'whiteAlpha.800'
    }
  };
};

var subtleStyle = function subtleStyle(ref) {
  var colors = ref.theme.colors;
  var color = ref.color;

  var _color = colors[color] && colors[color][200];

  var alphaColors = utils.generateAlphas(_color);
  var darkModebackgroundColor = alphaColors[300];
  return {
    light: {
      backgroundColor: utils.get(color, 100),
      color: utils.get(color, 800)
    },
    dark: {
      backgroundColor: darkModebackgroundColor,
      color: utils.get(color, 200)
    }
  };
};

var outlineStyle = function outlineStyle(ref) {
  var colors = ref.theme.colors;
  var color = ref.color;

  var _color = colors[color] && colors[color][200];

  var darkModeColor = utils.addOpacity(_color, 0.8);
  var boxShadowColor = colors[color] && colors[color][500];
  return {
    light: {
      boxShadow: 'inset 0 0 0px 1px ' + boxShadowColor,
      color: utils.get(color, 500)
    },
    dark: {
      boxShadow: 'inset 0 0 0px 1px ' + darkModeColor,
      color: darkModeColor
    }
  };
};

var variantProps = function variantProps(props) {
  var variant = props.variant;
  var colorMode = props.colorMode;

  switch (variant) {
    case 'solid':
      return solidStyle(props)[colorMode];

    case 'subtle':
      return subtleStyle(props)[colorMode];

    case 'outline':
      return outlineStyle(props)[colorMode];

    default:
      return {};
  }
};

var useBadgeStyle = function useBadgeStyle(props) {
  return variantProps(props);
};

exports.useBadgeStyle = useBadgeStyle;
