// @chakra-ui/vue/dist/cjs/CAlertDialog.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('./index-aaaef82b.js');
require('./chakra-ui-styled-system.esm-0de8acbf.js');
require('./internal-icons-971c36f2.js');
require('@emotion/css');
require('vue');
require('./color-mode-observer-121d1df0.js');
require('./anime.es-43f447cf.js');
require('./CBox.js');
require('./index-190bd03e.js');
require('./closebutton.styles-50adc572.js');
require('./CTransition.js');
var CModal = require('./CModal.js');
require('./CPortal.js');
require('./CNoSsr.js');
require('./props.types-2d254dfb.js');

/**
 * Hey! Welcome to @chakra-ui/vue AlertDialog
 *
 * AlertDialog component is used interrupt the user with
 * a mandatory confirmation or action.
 *
 * An alert dialog is a modal dialog that interrupts the user's
 * workflow to communicate an important message and acquire a response.
 * Examples include action confirmation prompts and error message confirmations.
 * The `alertdialog` role enables assistive technologies and browsers to
 * distinguish alert dialogs from other dialogs so they have the option
 * of giving alert dialogs special treatment, such as playing a system alert sound.
 *
 * @see Docs     https://vue.chakra-ui.com/alertdialog
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CAlertDialog/CAlertDialog.js
 * @see A11y     https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CAlertDialog/accessibility.md
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2/#alertdialog
 */

var formatIds = function formatIds(id) {
  return {
    content: "alert-dialog-" + id,
    header: "alert-dialog-" + id + "-label",
    body: "alert-dialog-" + id + "-desc"
  };
};
/**
 * CAlertDialog component
 *
 * The wrapper for alertdialog components. It provides context and state for the dialog.
 *
 * @extends CModal
 * @see Docs https://vue.chakra-ui.com/alertdialog
 */


var CAlertDialog = {
  name: 'CAlertDialog',
  functional: true,
  props: {
    isOpen: {
      type: Boolean,
      default: false
    },
    onClose: {
      type: Function,
      default: function _default() {
        return null;
      }
    },
    leastDestructiveRef: [utils.HTMLElement, Object, String, Function]
  },
  render: function render(h, ref) {
    var slots = ref.slots;
    var props = ref.props;
    var data = ref.data;
    return h(CModal.CModal, {
      props: {
        isOpen: props.isOpen,
        onClose: props.onClose,
        initialFocusRef: props.leastDestructiveRef,
        formatIds: formatIds
      },
      attrs: Object.assign({}, data.attrs, {
        'data-chakra-component': 'CAlertDialog'
      })
    }, slots().default);
  }
};
/**
 * CAlertDialogContent component
 *
 * The wrapper for the alert dialog's content.
 *
 * @extends CModalContent
 * @see Docs https://vue.chakra-ui.com/alertdialog
 */

var CAlertDialogContent = {
  name: 'CAlertDialogContent',
  functional: true,
  render: function render(h, ref) {
    var data = ref.data;
    var slots = ref.slots;
    return h(CModal.CModalContent, {
      attrs: Object.assign({}, {
        role: 'alertdialog'
      }, data.attrs, {
        'data-chakra-component': 'CAlertDialogContent'
      })
    }, slots().default);
  }
};
/**
 * CAlertDialogOverlay component
 *
 * The dimmed overlay behind the dialog.
 *
 * @extends CModalOverlay
 * @see Docs https://vue.chakra-ui.com/alertdialog
 */

var CAlertDialogOverlay = Object.assign({}, CModal.CModalOverlay, {
  name: 'CAlertDialogOverlay'
});
/**
 * CAlertDialogBody component
 *
 * Should contain the description announced by screen readers
 *
 * @extends CModalBody
 * @see Docs https://vue.chakra-ui.com/alertdialog
 */

var CAlertDialogBody = Object.assign({}, CModal.CModalBody, {
  name: 'CAlertDialogBody'
});
/**
 * CAlertDialogHeader component
 *
 * Should contain the title announced by screen readers
 *
 * @extends CModalHeader
 * @see Docs https://vue.chakra-ui.com/alertdialog
 */

var CAlertDialogHeader = Object.assign({}, CModal.CModalHeader, {
  name: 'CAlertDialogHeader'
});
/**
 * CAlertDialogFooter component
 *
 * Should contain the actions of the dialog
 *
 * @extends CModalFooter
 * @see Docs https://vue.chakra-ui.com/alertdialog
 */

var CAlertDialogFooter = Object.assign({}, CModal.CModalFooter, {
  name: 'CAlertDialogFooter'
});
/**
 * CAlertDialogCloseButton component
 *
 * The button that closes the dialog.
 *
 * @extends CModalCloseButton
 * @see Docs https://vue.chakra-ui.com/alertdialog
 */

var CAlertDialogCloseButton = Object.assign({}, CModal.CModalCloseButton, {
  name: 'CAlertDialogCloseButton'
});

exports.CAlertDialog = CAlertDialog;
exports.CAlertDialogBody = CAlertDialogBody;
exports.CAlertDialogCloseButton = CAlertDialogCloseButton;
exports.CAlertDialogContent = CAlertDialogContent;
exports.CAlertDialogFooter = CAlertDialogFooter;
exports.CAlertDialogHeader = CAlertDialogHeader;
exports.CAlertDialogOverlay = CAlertDialogOverlay;
