// @chakra-ui/vue/dist/cjs/CDivider.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('./index-aaaef82b.js');
require('./chakra-ui-styled-system.esm-0de8acbf.js');
require('@emotion/css');
require('vue');
require('./color-mode-observer-121d1df0.js');

/**
 * Hey! Welcome to @chakra-ui/vue Divider
 *
 * Dividers are used to display a thin horizontal or vertical line.
 *
 * @see Docs     https://vue.chakra-ui.com/divider
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CDivider/CDivider.js
 * @see A11y     https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CDivider/accessibility.md
 */
/**
 * CDivider component
 *
 * Creates a horizontal or vertical dividing rule between sibling
 * elements
 *
 * @see Docs https://vue.chakra-ui.com/divider
 */

var CDivider = {
  name: 'CDivider',
  mixins: [utils.createStyledAttrsMixin('CDivider')],
  props: {
    orientation: {
      type: String,
      default: 'horizontal'
    }
  },
  computed: {
    borderProps: function borderProps() {
      return this.orientation === 'vertical' ? {
        borderLeft: '0.0625rem solid',
        height: 'auto',
        mx: 2
      } : {
        borderBottom: '0.0625rem solid',
        width: 'auto',
        my: 2
      };
    },
    componentStyles: function componentStyles() {
      return Object.assign({}, {
        border: 0
      }, this.borderProps, {
        opacity: 0.6,
        borderColor: 'inherit'
      });
    }
  },
  render: function render(h) {
    return h('hr', {
      class: this.className,
      attrs: Object.assign({}, {
        'aria-orientation': this.orientation
      }, this.computedAttrs)
    });
  }
};

exports.default = CDivider;
