// @chakra-ui/vue/dist/cjs/CFormHelperText.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('./index-aaaef82b.js');
require('./chakra-ui-styled-system.esm-0de8acbf.js');
require('@emotion/css');
require('vue');
require('./color-mode-observer-121d1df0.js');

/**
 * Hey! Welcome to @chakra-ui/vue CFormHelperText
 *
 * Used to display helpful hints to the use on how to
 * use an input.
 *
 * @see Docs     https://vue.chakra-ui.com/formcontrol
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CFormHelperText/CFormHelperText.js
 * @see WAI      https://www.w3.org/WAI/tutorials/forms/
 */
/**
 * CFormHelperText component
 *
 * Used to display helpful hints to the use on how to
 * use an input.
 *
 * @extends CText
 * @see Docs https://vue.chakra-ui.com/formcontrol
 */

var CFormHelperText = {
  name: 'CFormHelperText',
  mixins: [utils.createStyledAttrsMixin('CFormHelperText')],
  inject: ['$useFormControl'],
  computed: {
    formControl: function formControl() {
      return this.$useFormControl(this.$props);
    },
    componentStyles: function componentStyles() {
      var color = {
        light: 'gray.500',
        dark: 'whiteAlpha.600'
      };
      return {
        mt: 2,
        color: color[this.colorMode],
        lineHeight: 'normal',
        fontSize: 'sm'
      };
    }
  },
  render: function render(h) {
    return h('p', {
      class: [this.className],
      attrs: Object.assign({}, {
        id: this.formControl.id ? this.formControl.id + "-help-text" : null
      }, this.computedAttrs),
      on: this.computedListeners
    }, this.$slots.default);
  }
};

exports.default = CFormHelperText;
