// @chakra-ui/vue/dist/cjs/CList.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('./index-aaaef82b.js');
require('./chakra-ui-styled-system.esm-0de8acbf.js');
require('./internal-icons-971c36f2.js');
require('@emotion/css');
require('vue');
require('./color-mode-observer-121d1df0.js');
var CIcon = require('./index-190bd03e.js');
var props_types = require('./props.types-2d254dfb.js');

/**
 * Hey! Welcome to @chakra-ui/vue List
 *
 * `CList` is used to display list items
 *
 * @see Docs     https://vue.chakra-ui.com/list
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CList/CList.js
 */
function objectWithoutProperties(obj, exclude) {
  var target = {};

  for (var k in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, k) && exclude.indexOf(k) === -1) target[k] = obj[k];
  }

  return target;
}
/**
 * CList component
 *
 * The list container element
 *
 * @extends CBox
 * @see Docs https://vue.chakra-ui.com/list
 */

var CList = {
  name: 'CList',
  mixins: [utils.createStyledAttrsMixin('CList')],
  props: {
    styleType: {
      type: String,
      default: 'none'
    },
    stylePos: {
      type: String,
      default: 'inside'
    },
    spacing: props_types.SNA
  },
  computed: {
    componentStyles: function componentStyles() {
      return {
        listStyleType: this.styleType,
        listStylePosition: this.stylePos
      };
    }
  },
  render: function render(h) {
    var this$1 = this;
    var children = this.$slots.default;

    if (!utils.isDef(children)) {
      console.error('[Chakra-ui: List]: List component expects at east one child');
      return null;
    }

    var validChildren = utils.cleanChildren(children);
    var clones = validChildren.map(function (vnode, index) {
      var isLast = index + 1 === validChildren.length;

      if (isLast) {
        return vnode;
      }

      var clone = utils.cloneVNodeElement(vnode, {
        attrs: {
          mb: this$1.spacing
        }
      }, h);
      return clone;
    });
    return h('ul', {
      class: [this.className],
      attrs: this.computedAttrs,
      on: this.computedListeners
    }, clones);
  }
};
/**
 * CListItem component
 *
 * The list item element
 *
 * @extends CPseudoBox
 * @see Docs https://vue.chakra-ui.com/list
 */

var CListItem = {
  name: 'CListItem',
  mixins: [utils.createStyledAttrsMixin('CListItem')],
  props: {
    spacing: props_types.SNA
  },
  computed: {
    componentStyles: function componentStyles() {
      return {
        mb: this.spacing
      };
    }
  },
  render: function render(h) {
    return h('li', {
      class: [this.className],
      attrs: this.computedAttrs
    }, this.$slots.default);
  }
};
/**
 * CListIcon component
 *
 * The list item icon element
 *
 * @extends CIcon
 * @see Docs https://vue.chakra-ui.com/list
 */

var CListIcon = {
  name: 'CListIcon',
  functional: true,
  props: {
    icon: String
  },
  render: function render(h, ref) {
    var props = ref.props;
    var data = ref.data;
    var rest$1 = objectWithoutProperties(ref, ["props", "data"]);
    var rest = rest$1;
    return h(CIcon.CIcon, Object.assign({}, rest, {
      props: {
        name: props.icon
      },
      attrs: Object.assign({}, {
        mr: 2
      }, data.attrs, {
        'data-chakra-component': 'CListIcon'
      })
    }));
  }
};

exports.CListIcon = CListIcon;
exports.CListItem = CListItem;
exports.default = CList;
