// @chakra-ui/vue/dist/cjs/CNoSsr.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Hey! Welcome to @chakra-ui/vue NoSsr
 *
 * This component is used to defer the rendering of it's children to the client-side.
 *
 * This component is a modification of the fine work of @egoist
 * @see https://github.com/egoist/vue-client-only/blob/master/src/index.js
 */
var NoSsr = {
  name: 'NoSsr',
  functional: true,
  props: {
    placeholder: String,
    placeholderTag: {
      type: String,
      default: 'div'
    }
  },
  render: function render(h, ref) {
    var parent = ref.parent;
    var slots = ref.slots;
    var props = ref.props;
    var ref$1 = slots();
    var defaultSlot = ref$1.default;
    if (defaultSlot === void 0) defaultSlot = [];
    var placeholderSlot = ref$1.placeholder;

    if (parent._isMounted) {
      return defaultSlot;
    }

    parent.$once('hook:mounted', function () {
      parent.$forceUpdate();
    });

    if (props.placeholderTag && (props.placeholder || placeholderSlot)) {
      return h(props.placeholderTag, {
        class: ['client-only-placeholder'],
        attrs: {
          'data-chakra-component': 'CNoSsr'
        }
      }, props.placeholder || placeholderSlot);
    } // Return a placeholder element for each child in the default slot
    // Or if no children return a single placeholder


    return defaultSlot.length > 0 ? defaultSlot.map(function () {
      return h(false);
    }) : h(false);
  }
};

exports.default = NoSsr;
