// @chakra-ui/vue/dist/cjs/CStat.js

/*
 * ! @chakra-ui/vue v0.13.1 by Jonathan Bakebwa @codebender828
 * MIT License
 * https://github.com/chakra-ui/chakra-ui-vue
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('./index-aaaef82b.js');
require('./chakra-ui-styled-system.esm-0de8acbf.js');
require('./internal-icons-971c36f2.js');
require('@emotion/css');
require('vue');
require('./color-mode-observer-121d1df0.js');
var CBox = require('./CBox.js');
var CIcon = require('./index-190bd03e.js');
require('./props.types-2d254dfb.js');
var CFlex = require('./index-6708ec7a.js');
require('./text.utils-e344d432.js');
var CText = require('./CText.js');

/**
 * Hey! Welcome to @chakra-ui/vue Stat
 *
 * The Stat component is used to display a single statistic.
 *
 * @see Docs     https://vue.chakra-ui.com/stat
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CStat/CStat.js
 */
function objectWithoutProperties(obj, exclude) {
  var target = {};

  for (var k in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, k) && exclude.indexOf(k) === -1) target[k] = obj[k];
  }

  return target;
}
/**
 * Stat Arrow options
 */

var arrowOptions = {
  increase: {
    name: 'triangle-up',
    color: 'green.400'
  },
  decrease: {
    name: 'triangle-down',
    color: 'red.400'
  }
};
/**
 * CStat component
 *
 * Stat wrapper component for it's children
 *
 * @extends CBox
 * @see Docs https://vue.chakra-ui.com/stat
 */

var CStat = {
  name: 'CStat',
  functional: true,
  render: function render(h, ref) {
    var props = ref.props;
    var slots = ref.slots;
    var data = ref.data;
    var rest$1 = objectWithoutProperties(ref, ["props", "slots", "data"]);
    var rest = rest$1;
    var children = utils.cleanChildren(slots().default);
    return h(CBox.default, Object.assign({}, rest, {
      props: {
        as: props.as
      },
      attrs: Object.assign({}, {
        flex: 1,
        pr: 4,
        position: 'relative'
      }, data.attrs || {}, {
        'data-chakra-component': 'CStat'
      })
    }), children);
  }
};
/**
 * CStatGroup component
 *
 * group component for stat
 *
 * @extends CFlex
 * @see Docs https://vue.chakra-ui.com/stat
 */

var CStatGroup = {
  name: 'CStatGroup',
  functional: true,
  props: CFlex.CFlex.props,
  render: function render(h, ref) {
    var props = ref.props;
    var slots = ref.slots;
    var data = ref.data;
    var rest$1 = objectWithoutProperties(ref, ["props", "slots", "data"]);
    var rest = rest$1;
    var children = utils.cleanChildren(slots().default);
    return h(CFlex.CFlex, Object.assign({}, rest, {
      props: Object.assign({}, {
        flexWrap: 'wrap',
        justifyContent: 'space-around',
        alignItems: 'flex-start'
      }, utils.pickProperty(props)),
      attrs: Object.assign({}, data.attrs || {}, {
        'data-chakra-component': 'CStatGroup'
      })
    }), children);
  }
};
/**
 * CStatArrow component
 *
 * Arrow component for stat
 *
 * @extends CIcon
 * @see Docs https://vue.chakra-ui.com/stat
 */

var CStatArrow = {
  name: 'CStatArrow',
  functional: true,
  props: Object.assign({}, CIcon.CIcon.props, {
    type: {
      type: String,
      default: 'increase'
    }
  }),
  render: function render(h, ref) {
    var props = ref.props;
    var slots = ref.slots;
    var data = ref.data;
    var rest$1 = objectWithoutProperties(ref, ["props", "slots", "data"]);
    var rest = rest$1;
    return h(CIcon.CIcon, Object.assign({}, rest, {
      props: Object.assign({}, {
        size: '14px'
      }, arrowOptions[props.type], utils.pickProperty(props)),
      attrs: Object.assign({}, {
        mr: 1,
        verticalAlign: 'middle'
      }, data.attrs || {}, {
        color: arrowOptions[props.type].color,
        'data-chakra-component': 'CStatArrow'
      })
    }));
  }
};
/**
 * CStatNumber component
 *
 * number component for stat
 *
 * @extends CIcon
 * @see Docs https://vue.chakra-ui.com/stat
 */

var CStatNumber = {
  name: 'CStatNumber',
  functional: true,
  props: CText.default.props,
  render: function render(h, ref) {
    var props = ref.props;
    var slots = ref.slots;
    var data = ref.data;
    var rest$1 = objectWithoutProperties(ref, ["props", "slots", "data"]);
    var rest = rest$1;
    return h(CText.default, Object.assign({}, rest, {
      props: utils.pickProperty(props),
      attrs: Object.assign({}, {
        fontSize: '2xl',
        verticalAlign: 'baseline',
        fontWeight: 'semibold'
      }, data.attrs || {}, {
        'data-chakra-component': 'CStatNumber'
      })
    }), slots().default);
  }
};
/**
 * CStatHelperText component
 *
 * text component for stat
 *
 * @extends CText
 * @see Docs https://vue.chakra-ui.com/stat
 */

var CStatHelperText = {
  name: 'CStatHelperText',
  functional: true,
  props: CText.default.props,
  render: function render(h, ref) {
    var props = ref.props;
    var slots = ref.slots;
    var data = ref.data;
    var rest$1 = objectWithoutProperties(ref, ["props", "slots", "data"]);
    var rest = rest$1;
    return h(CText.default, Object.assign({}, rest, {
      props: utils.pickProperty(props),
      attrs: Object.assign({}, {
        fontSize: 'sm',
        opacity: 0.8,
        mb: 2
      }, data.attrs || {}, {
        'data-chakra-component': 'CStatHelperText'
      })
    }), slots().default);
  }
};
/**
 * CStatLabel component
 *
 * label text component for stat
 *
 * @extends CText
 * @see Docs https://vue.chakra-ui.com/stat
 */

var CStatLabel = {
  name: 'CStatLabel',
  functional: true,
  props: CText.default.props,
  render: function render(h, ref) {
    var props = ref.props;
    var slots = ref.slots;
    var data = ref.data;
    var rest$1 = objectWithoutProperties(ref, ["props", "slots", "data"]);
    var rest = rest$1;
    return h(CText.default, Object.assign({}, rest, {
      props: utils.pickProperty(props),
      attrs: Object.assign({}, {
        fontWeight: 'medium',
        fontSize: 'sm'
      }, data.attrs || {}, {
        'data-chakra-component': 'CStatLabel'
      })
    }), slots().default);
  }
};

exports.CStat = CStat;
exports.CStatArrow = CStatArrow;
exports.CStatGroup = CStatGroup;
exports.CStatHelperText = CStatHelperText;
exports.CStatLabel = CStatLabel;
exports.CStatNumber = CStatNumber;
